### DISCUSSION TITLE
Adding support for custom esp32 device

### DISCUSSION THREAD

**Discussion by 0rion12** [Koenkk/zigbee2mqtt#23142] [Devices]
Hello, 

My goal is to successfully create a custom zigbee device from an esp32-c6 and to connect and control it to my home automation system.

I think I have gone through all the discussions that deal with the same subject, but either the problem discussed cannot be applied to my issue, or the issue is similar, but no one has provided an answer. So let me reopen a discussion.

To summarize, my home automation installation is as follows:
A raspberry pi 4 on which Home assistant and the zigbee2mqtt extension are installed. I use SONOFF_Zigbee_3.0_USB_Dongle_Plus_V2 usb dongle for zigbee connectivity.

I already use commercial Zigbee equipment (including bulbs and switches from Ikea and Lidl). But I want to be able to make my own custom zigbee devices. For this, I bought an esp32-c6 card to do tests.

After a lot of testing and searching on the forums, I managed to get my esp32 to join my zigbee network, to be recognized as a lamp and to have its model and manufacturer recognized. Here is the configuration of my device as detected under z2m:

![image](https://github.com/Koenkk/zigbee2mqtt/assets/61796302/42c1ba09-ae8a-4407-a5a5-6590f8adc401)

Obviously, as I put a model and a manufacturer which is not that of a commercial device, it says that it is not supported.
Here is the list of available clusters (this corresponds well to the clusters that I defined in the code of my esp32):

<img width="293" alt="image" src="https://github.com/Koenkk/zigbee2mqtt/assets/61796302/bbb0ee31-9986-4878-a49d-8e45d8546536">

The problem is that when I go to the "exposes" tab, the only command I have is an on/off button, which works well (this changes the state of the LED on the esp32 by pressing the button). But I don't see any possibility of changing the brightness which is nevertheless one of the clusters. 

<img width="685" alt="image" src="https://github.com/Koenkk/zigbee2mqtt/assets/61796302/5a0d0209-b497-4c76-a850-10ba7cfd5f16">

Under the "dev console" tab, I can read the value of this cluster, which means that it is indeed accessible.
By searching the forums and the z2m documentation, I understood that I had to add an external converter. So in the "/homeassistant/zigbee2mqtt" folder I added a file "ESPRESSIV.js" (I understood that the name of the file had to correspond to the manufacturer. Here is its content:
```
import {Definition} from '../lib/types';
import {light} from '../lib/modernExtend';

const definitions: Definition[] = [
    {
        zigbeeModel: ['TEST1'],
        model: 'TEST1',
        vendor: 'ESPRESSIV',
        description: 'WeMo smart LED bulb',
        extend: [light()],
    },
];

export default definitions;
module.exports = definitions;
```
Then, I modified the file "/homeassistant/zigbee2mqtt/configuration.yaml" to add these lines:

```
data_path : /homeassistant/zigbee2mqtt
homeassistant: true
external_converters:
  - ESPRESSIV.js
```

Then I restarted Home Assistant, but nothing changed for my device. I don't have any other controllable clusters in the "Exposes" tab.
I also tried removing the device from the zigbee network and adding it back to see if having the converter already present at the time of joining could improve things, but nothing changed.

I tried different scripts for the  "ESPRESSIV.js" file, using the old and new way of defining clusters. I tried using the "light" entity only or by specifying the different clusters that I wanted 1 by 1, but the result was always the same, I have the impression that the script is never taken into account and z2m does not make the link between my device and this script.

I don't know at what point in the process it gets stuck. Is the code I put in the “configuration.yaml” and “ESPRESSIV.js” files correct? Is the process I followed correct? Or is there something I should have done differently?

I hope that someone could help me.

Thank you so much.

**Comment by elgerg** [comment]
Hi,

Sorry if I'm really late to the game on this one but the later versions of Z2M support the ability to generate configs.

Here is the one I made from the Arduino IDE but for the temp sensor:

![image](https://github.com/user-attachments/assets/48ae5d96-cc78-4c4b-8c8d-0f28978f149a)

![image](https://github.com/user-attachments/assets/2e004312-69c1-46af-825a-c6ed4d777807)

Hope that helps!

**Reply by 0rion12** [reply]
Hi @elgerg,
Thank you so much ! By putting the generated code in my ESPRESSIV.js file, deleting the device, restarting Z2M and doing a new pairing, it works thank you!

**Comment by elgerg** [comment]
Oh wow, I'd completely forgotten about this. Glad I could help!

**Comment by czagaadam** [comment]
hi @0rion12 

I'm working on a similiar project. Try to pair an ESP32 H2 device with a zigbee network.
The network is formed by RPI 3B+ with a sonoff dongle (I'm not user this part is okay but zigbee2mqtt can be launced successfully).
The esp_zb_light example is flashed on H2 but I always getting
I (545840) ESP_ZB_ON_OFF_LIGHT: Network steering was not successful (status: ESP_FAIL) error.
You mentioned in your first comment you managed to solve this. 
Can you please share which example you have used and what was modified in it.
Are you using the network key, pan id and extended pan id ?

**Reply by 0rion12** [reply]
Hi @czagaadam,

If the goal is just to connect the esp32 to your zigbee network, I just tried again to be sure i'm not saying wrong things and I managed to connect my esp32-C6 to my zigbee2mqtt integration on home assistant. Here are the steps:

I downloaded the latest version of esp-zigbee-sdk: https://github.com/espressif/esp-zigbee-sdk.

I opened the HA_on_off_light folder located in the examples folder with vscode. I installed the latest version of esp-idf (5.4.0). I did not change anything in the project configuration or in the code.

I cleared the esp memory via the option found in the advanced options of esp-idf. Then I flashed my esp32 with the code.

Once flashed, here are my logs:

I (23) boot: ESP-IDF v5.4 2nd stage bootloader
I (23) boot: compile time Feb 21 2025 21:41:59
I (24) boot: chip revision: v0.1
I (24) boot: efuse block revision: v0.3
I (26) boot.esp32c6: SPI Speed      : 80MHz
I (30) boot.esp32c6: SPI Mode       : DIO
I (34) boot.esp32c6: SPI Flash Size : 2MB
I (38) boot: Enabling RNG early entropy source...
I (42) boot: Partition Table:
I (45) boot: ## Label            Usage          Type ST Offset   Length
I (51) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (58) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (64) boot:  2 factory          factory app      00 00 00010000 000e1000
I (71) boot:  3 zb_storage       Unknown data     01 81 000f1000 00004000
I (77) boot:  4 zb_fct           Unknown data     01 81 000f5000 00000400
I (84) boot: End of partition table
I (87) esp_image: segment 0: paddr=00010020 vaddr=42060020 size=0eb60h ( 60256) map
I (106) esp_image: segment 1: paddr=0001eb88 vaddr=40800000 size=01490h (  5264) load
I (108) esp_image: segment 2: paddr=00020020 vaddr=42000020 size=58104h (360708) map
I (176) esp_image: segment 3: paddr=0007812c vaddr=40801490 size=0daf4h ( 56052) load
I (189) esp_image: segment 4: paddr=00085c28 vaddr=4080ef90 size=02390h (  9104) load
I (195) boot: Loaded app from partition at offset 0x10000
I (195) boot: Disabling RNG early entropy source...
I (206) cpu_start: Unicore app
I (215) cpu_start: Pro cpu start user code
I (215) cpu_start: cpu freq: 160000000 Hz
I (215) app_init: Application information:
I (216) app_init: Project name:     on_off_light_bulb
I (220) app_init: App version:      1
I (223) app_init: Compile time:     Feb 21 2025 21:41:34
I (228) app_init: ELF file SHA256:  31265b48b...
I (233) app_init: ESP-IDF:          v5.4
I (236) efuse_init: Min chip rev:     v0.0
I (240) efuse_init: Max chip rev:     v0.99 
I (244) efuse_init: Chip rev:         v0.1
I (248) heap_init: Initializing. RAM available for dynamic allocation:
I (254) heap_init: At 40815620 len 00066FF0 (411 KiB): RAM
I (259) heap_init: At 4087C610 len 00002F54 (11 KiB): RAM
I (264) heap_init: At 50000000 len 00003FE8 (15 KiB): RTCRAM
I (271) spi_flash: detected chip: generic
I (274) spi_flash: flash io: dio
W (276) spi_flash: Detected size(4096k) larger than the size in the binary image header(2048k). Using the size in the binary image header.
I (289) sleep_gpio: Configure to isolate all GPIO pins in sleep state
I (295) sleep_gpio: Enable automatic switching of GPIO sleep configuration
I (302) coexist: coex firmware version: 49a8cdc
I (306) coexist: coexist rom version 5b8dcfa
I (310) main_task: Started on CPU0
I (310) main_task: Calling app_main()
I (320) phy_init: phy_version 320,348a293,Sep  3 2024,16:33:12
I (390) phy: libbtbb version: 04952fd, Sep  3 2024, 16:33:30
I (390) main_task: Returned from app_main()
I (400) ESP_ZB_ON_OFF_LIGHT: ZDO signal: ZDO Config Ready (0x17), status: ESP_FAIL
I (400) ESP_ZB_ON_OFF_LIGHT: Initialize Zigbee stack
I (400) ESP_ZB_ON_OFF_LIGHT: Deferred driver initialization successful
I (400) ESP_ZB_ON_OFF_LIGHT: Device started up in factory-reset mode
I (410) ESP_ZB_ON_OFF_LIGHT: Start network steering
I (2660) ESP_ZB_ON_OFF_LIGHT: Network steering was not successful (status: ESP_FAIL)
I (5910) ESP_ZB_ON_OFF_LIGHT: Network steering was not successful (status: ESP_FAIL)
I (9160) ESP_ZB_ON_OFF_LIGHT: Network steering was not successful (status: ESP_FAIL)

So here I think I have the same behavior than you. So I went to my zigbee2mqtt extension and clicked on the "Permit join" button to allow new devices to join. Then after few seconds, the device joined the network and appeared on z2m integration : 

I (18910) ESP_ZB_ON_OFF_LIGHT: Network steering was not successful (status: ESP_FAIL)
I (22160) ESP_ZB_ON_OFF_LIGHT: Network steering was not successful (status: ESP_FAIL)
I (25410) ESP_ZB_ON_OFF_LIGHT: Network steering was not successful (status: ESP_FAIL)
I (28660) ESP_ZB_ON_OFF_LIGHT: Network steering was not successful (status: ESP_FAIL)
I (32610) ESP_ZB_ON_OFF_LIGHT: Joined network successfully (Extended PAN ID: c1:85:9d:87:e8:17:80:f9, PAN ID: 0x6793, Channel:11, Short Address: 0xc1b2)
W (33790) ESP_ZB_ON_OFF_LIGHT: Receive Zigbee action(0x1005) callback
I (69320) ESP_ZB_ON_OFF_LIGHT: Received message: endpoint(10), cluster(0x6), attribute(0x0), data size(1)

![image](https://github.com/user-attachments/assets/8c54f969-f5c6-4d2a-9d8a-327bb5b6ef63)

If you have done all of this and you can not make your esp32 to join the network, a piece of advice i can give you is to bring the esp32 a few centimeters away from the zigbee antenna of your raspberry pi during the join sequence. Indeed, i have already had several errors for join and this could be solved by bringing the esp32 closer to the antenna significantly.

If you manage to get it to work like this or have any other questions, feel free to comment.

**Reply by czagaadam** [reply]
hi @0rion12 
Thanks for your feedback and for your efforts.
Finally I found a soultion, I changed the device name and type to xiaomi and some existing device so now it is working (I found this in another thread in github)

**Comment by Moralez1980** [comment]
@Orion12 Could you share your ESP32C6 code ?
